---
output:
  html_document:
    theme: spacelab
---

<a href="https://jhubiostatistics.shinyapps.io/recount/"><img src="https://raw.githubusercontent.com/leekgroup/recount-website/master/website/recount.png" align="center"></a>

Example analyses with the [recount](https://github.com/leekgroup/recount) R package using the [recount project](https://jhubiostatistics.shinyapps.io/recount/).

* [Meta analysis](example_meta/meta_analysis.html) ([PDF](example_meta/meta_analysis.pdf))
* [Differential expression at the gene level](example_de/recount_SRP019936.html) ([PDF](example_de/recount_SRP019936.pdf))
* [Differential expression at the exon level](example_de/recount_SRP032789.html) ([PDF](example_de/recount_SRP032789.pdf))
* [Annotation-agnostic differential expression](example_de/recount_DER_SRP032789.html) ([PDF](example_de/recount_DER_SRP032789.pdf))
* [GTEx counts comparison]() ([PDF]())


# References

```{r 'setup', echo = FALSE, warning = FALSE, message = FALSE}
## Bib setup
library('knitcitations')
library('BiocStyle')

## Load knitcitations with a clean bibliography
cleanbib()
cite_options(hyperlink = 'to.doc', citation_format = 'text', style = 'html')

## Write bibliography information
bibs <- c(
    BiocParallel = citation('BiocParallel'),
    BiocStyle = citation('BiocStyle'),
    derfinder = citation('derfinder')[1],
    devtools = citation('devtools'),
    dplyr = citation('dplyr'),
    edgeR = citation('edgeR')[1],
    ffpe = citation('ffpe'),
    IHW = citation('IHW'),
    knitcitations = citation('knitcitations'),
    magrittr = citation('magrittr'),
    matrixStats = citation('matrixStats'),
    qvalue = citation('qvalue'),
    R = citation(),
    recount = citation('recount'),
    rmarkdown = citation('rmarkdown'),
    RSkittleBrewer = citation('RSkittleBrewer'),
    SummarizedExperiment = citation('SummarizedExperiment'),
    topGO = citation('topGO'),
    voom = RefManageR::BibEntry('article', key = 'voom', author = 'CW Law and Y Chen and W Shi and GK Smyth', year = '2014', title = 'Voom: precision weights unlock linear model analysis tools for RNA-seq read counts', journal = 'Genome Biology', volume = '15', pages = 'R29')
)

write.bibtex(bibs, file = 'index.bib')
bib <- read.bibtex('index.bib')

## Assign short names
names(bib) <- names(bibs)
```

The analyses were made possible thanks to:

* R `r citep(bib[['R']])`
* `r Biocpkg('BiocParallel')` `r citep(bib[['BiocParallel']])`
* `r Biocpkg('BiocStyle')` `r citep(bib[['BiocStyle']])`
* `r Biocpkg('derfinder')` `r citep(bib[['derfinder']])`
* `r CRANpkg('devtools')` `r citep(bib[['devtools']])`
* `r CRANpkg('dplyr')` `r citep(bib[['dplyr']])`
* `r Biocpkg('edgeR')` `r citep(bib[['edgeR']])`
* `r Biocpkg('ffpe')` `r citep(bib[['ffpe']])`
* `r Biocpkg('IHW')` `r citep(bib[['IHW']])`
* `r CRANpkg('knitcitations')` `r citep(bib[['knitcitations']])`
* `r CRANpkg('magrittr')` `r citep(bib[['magrittr']])`
* `r CRANpkg('matrixStats')` `r citep(bib[['matrixStats']])`
* `r Biocpkg('qvalue')` `r citep(bib[['qvalue']])`
* `r Githubpkg('leekgroup/recount')` `r citep(bib[['recount']])`
* `r CRANpkg('rmarkdown')` `r citep(bib[['rmarkdown']])`
* `r Githubpkg('alyssafrazee/RSkittleBrewer')` `r citep(bib[['RSkittleBrewer']])`
* `r Biocpkg('SummarizedExperiment')` `r citep(bib[['SummarizedExperiment']])`
* `r Biocpkg('topGO')` `r citep(bib[['topGO']])`
* `r Biocpkg('limma')` `r citep(bib[['voom']])`

[Bibliography file](index.bib)

```{r bibliography, results='asis', echo=FALSE, warning = FALSE}
## Print bibliography
bibliography()
```