---
title: 'Recount Meta-Analysis'
author: 'Jeff Leek and Margaret Taub'
date: 'June 7, 2016'
output:
  BiocStyle::html_document:
    toc: true
  pdf_document:
    toc: true
---

```{r 'setup', echo = FALSE, warning = FALSE, message = FALSE}
timestart <- Sys.time()

## Bib setup
library('knitcitations')

## Load knitcitations with a clean bibliography
cleanbib()
cite_options(hyperlink = 'to.doc', citation_format = 'text', style = 'html')

## Write bibliography information
bibs <- c(
    BiocStyle = citation('BiocStyle'),
    derfinder = citation('derfinder')[1],
    devtools = citation('devtools'),
    dplyr = citation('dplyr'),
    edgeR = citation('edgeR')[1],
    ffpe = citation('ffpe'),
    knitcitations = citation('knitcitations'),
    magrittr = citation('magrittr'),
    R = citation(),
    recount = citation('recount'),
    rmarkdown = citation('rmarkdown'),
    RSkittleBrewer = citation('RSkittleBrewer'),
    SummarizedExperiment = citation('SummarizedExperiment'),
    voom = RefManageR::BibEntry('article', key = 'voom', author = 'CW Law and Y Chen and W Shi and GK Smyth', year = '2014', title = 'Voom: precision weights unlock linear model analysis tools for RNA-seq read counts', journal = 'Genome Biology', volume = '15', pages = 'R29'),
    colon1 = RefManageR::BibEntry('article', key = 'colon1', author = 'Seon-Kyu Kim and Seon-Young Kim and Jeong-Hwan Kim and Seon Ae Roh and Dong-Hyung Cho and Yong Sung Kim and Jin Cheon Kim', year = '2014', title = 'A nineteen gene-based risk score classifier predicts prognosis of colorectal cancer patients ', journal = 'Molecular Oncology', volume = '8', pages = '1653 - 1666'),
    colon2 = RefManageR::BibEntry('article', key = 'colon2', author = 'Yael Haberman AND Timothy L. Tickle AND Phillip J. Dexheimer AND Mi-Ok Kim AND Dora Tang AND Rebekah Karns AND Robert N. Baldassano AND Joshua D. Noe AND Joel Rosh AND James Markowitz AND Melvin B. Heyman AND Anne M. Griffiths AND Wallace V. Crandall AND David R. Mack AND Susan S. Baker AND Curtis Huttenhower AND David J. Keljo AND Jeffrey S. Hyams AND Subra Kugathasan AND Thomas D. Walters AND Bruce Aronow AND Ramnik J. Xavier AND Dirk Gevers AND Lee A. Denson', year = '2014', title = 'Pediatric Crohn disease patients exhibit specific ileal transcriptome and microbiome signature', journal = 'The Journal of Clinical Investigation', volume = '124', pages = '3617-3633')
)

write.bibtex(bibs,
    file = 'meta_analysis.bib')
bib <- read.bibtex('meta_analysis.bib')

## Assign short names
names(bib) <- names(bibs)
```

# Analysis
To illustrate the ease of combining data from multiple projects included in `recount` as part of a cross-study meta-analysis, we present here a cross-tissue differential expression (DE) analysis comparing gene expression between colon and whole blood. We perform an initial analysis on samples from a variety of studies found in the `recount` project and then compare the results we obtain to those from analysis of GTEx data in the same tissues.



## Load libraries we will need

```{r loadpackages, message = FALSE, warning = FALSE}
library('dplyr')
library('recount')
library('magrittr')
library('limma')
library('edgeR')
library('ffpe')
library('RSkittleBrewer')
library('SummarizedExperiment')
library('devtools')
trop <- RSkittleBrewer::RSkittleBrewer('tropical')
```

## Get data sets

As an initial analysis, colon samples labeled as controls were taken from studies `SRP029880` (a study of colorectal cancer `r citep(bib[["colon1"]])`, n=19) and `SRP042228` (a study of Crohn's disease `r citep(bib[['colon2']])`, n=41). Whole blood samples labeled as controls were taken from `SRP059039` (a study of virus-caused diarrhea, unpublished, n=24), `SRP059172` (a study of blood biomarkers for brucellosis, unpublished, n=47) and `SRP062966` (a study of lupus, unpublished, n=18). 

We start by downloading `.Rdata` files containing `SummarizedExperiment` objects for each of these studies, which contain RNA-seq counts at the gene level for each individual in the study, as well as accompanying sample-specific phenotype data.


```{r selectprojects}
colon_proj <- c('SRP029880', 'SRP042228')
if(any(!file.exists(file.path(colon_proj, 'rse_gene.Rdata')))) {
    sapply(colon_proj, download_study)
}

blood_proj <- c('SRP059039', 'SRP059172', 'SRP062966')
if(any(!file.exists(file.path(blood_proj, 'rse_gene.Rdata')))) {
    sapply(blood_proj, download_study)
}

proj <- c(colon_proj,blood_proj)
```

## Load the data

Once the data has been downloaded, we load all five data sets into `R` and calculate the number of genes (number of rows) and samples (number of columns) for each data set.

```{r loaddata}
dat <- lapply(proj, function(x) { 
    load(file.path(x, 'rse_gene.Rdata'))
    return(rse_gene)
})
proj
sapply(dat, dim)
```


## Load the metadata file

Since the data contained in the `SummarizedExperiment` objects for each study are somewhat limited, we also load a data frame containing metadata for each sample, including Gene Expression Omnibus (GEO) sample identifiers. We will use this in order to identify and extract the control subjects from each study for analysis.

```{r loadmeta}
metadata <- all_metadata('sra')
```

## Get additional GEI data for these samples

Now we go through and collect GEO information for the samples. We label them with their respective tissue and identify which samples are supposed to be normal controls, with labels varying from data set to data set. We also label the data sets with their tissue of origin.

```{r getgeoinfo, warning = FALSE}
if(!file.exists('charvec.Rdata')) {
    charvec <- vector('list', 5)
    dir.create('geoinfo', showWarnings = FALSE)
    for(i in 1:5){
      index <- match(colData(dat[[i]])$run, metadata$run)
      colData(dat[[i]])$geo <- metadata$geo_accession[index]
      info <- sapply(colData(dat[[i]])$geo, geo_info, destdir = 'geoinfo')
      charvec[[i]] <- sapply(info, geo_characteristics)
    }
    save(charvec, file = 'charvec.Rdata')
} else {
    load('charvec.Rdata')
}

## first data set - normals called 'normal-looking surrounding colonic epithelium'
colData(dat[[1]])$normal <- grepl('normal', unlist(charvec[1])[(1:54) * 2 - 1])
colData(dat[[1]])$tissue <- 'colon'

## second data set - normals called 'not ibd'
colData(dat[[2]])$normal <- grepl('not ibd', tolower(unlist(charvec[[2]][5, ])))
colData(dat[[2]])$tissue <- 'colon'

## third data set  - normals called 'Control'
colData(dat[[3]])$normal <- grepl('Control', unlist(charvec[[3]][2, ]))
colData(dat[[3]])$tissue <- 'blood'

## fourth data set  - normals called 'Control'
colData(dat[[4]])$normal <- grepl('Control', unlist(charvec[[4]][1, ]))
colData(dat[[4]])$tissue <- 'blood'

## fifth data set - normals called 'healthy'
colData(dat[[5]])$normal <- grepl('healthy', unlist(charvec[[5]][1, ]))
colData(dat[[5]])$tissue <- 'blood'
```


## Merge the data sets

We merge the data sets into one ranged `SummarizedExperiment`.

```{r mergedata}
mdat <- do.call(cbind, dat)
```


## Subset to normal control individuals

Using the label determined above, we find out how many samples are normal in each study and subset to just the normal samples for further analysis. 

```{r selectnormals}
table(colData(mdat)$normal, colData(mdat)$project)
ndat <- mdat[, colData(mdat)$normal]
```


## Analysis comparing blood to colon

Here we do a differential expression analysis comparing blood to colon using `voom` `r citep(bib[['voom']])`. We start by normalizing the counts and then consider only the genes that have an average normalized count of at least 5 across the data set. 

```{r 'bloodvscolon'}
ndat <- scale_counts(ndat)
ndat_counts <- assays(ndat)[[1]]
keep <- rowMeans(ndat_counts) > 5
ndat_counts = ndat_counts[keep, ]
design <- model.matrix(~colData(ndat)$tissue)
dge <- DGEList(counts = ndat_counts)
dge <- calcNormFactors(dge)
v <- voom(dge, design, plot=FALSE)
fit <- lmFit(v, design)
fit <- eBayes(fit)
topTable(fit)
```


### GTEx analysis: comparing same tissues

In order to have something to compare to the results generated on the data sets selected above, we do a differential expression analysis comparing blood to colon samples from GTEx, project `SRP012682`. We carry out the same steps as above for the analysis, but in this case we also control for batch as included in the GTEx metadata.

```{r gtexanalysis}
## Download the GTEx data
if(!file.exists(file.path('SRP012682', 'rse_gene.Rdata'))) {
    download_study('SRP012682')
}
load(file.path('SRP012682', 'rse_gene.Rdata'))

## Download GTEx metadata
gtex_metadata <- all_metadata('gtex')

## Subset to samples from tissues of interest
gtex_blood <- rse_gene[, subset(gtex_metadata, smtsd == 'Whole Blood')$run]
colData(gtex_blood)$tissue <- 'wholeblood'
gtex_colon <- rse_gene[, subset(gtex_metadata, smts == 'Colon')$run]
colData(gtex_colon)$tissue <- 'colon'

## Combine tissue subsets and include batch label
gtex_both <- do.call(cbind, list(gtex_blood, gtex_colon))
colData(gtex_both)$batch <- gtex_metadata[match(colData(gtex_both)$run,
    gtex_metadata$run), 'smgebtch']

## Scale and filter data to the same genes kept in initial analysis
gtex_both <- scale_counts(gtex_both)
gtex_both_counts <- assays(gtex_both)[[1]]
gtex_both_counts <- gtex_both_counts[keep, ]

## Carry out differential expression analysis
design_gtex <- model.matrix(~colData(gtex_both)$tissue +
    colData(gtex_both)$batch)
dge_gtex <- DGEList(counts = gtex_both_counts)
dge_gtex <- calcNormFactors(dge_gtex)
v_gtex <- voom(dge_gtex, design_gtex, plot=FALSE)
fit_gtex <- lmFit(v_gtex, design_gtex)
fit_gtex <- eBayes(fit_gtex)
topTable(fit_gtex, coef = 2)
```


### GTEx analysis: comparing different tissues

To have a way of gauguing how much overlap we expect to see between the results from the five combined data sets and those from GTEx data, we also carry out a differential analysis using GTEx data comparing blood to lung. We expect to see much less concordance between our initial results and the GTEx blood-lung comparison than we see between our initial results and the GTEx blood-colon comparison. 


```{r gtexanalysislung}
## Subset lung data
gtex_lung <- rse_gene[, subset(gtex_metadata, smts=='Lung')$run]
colData(gtex_lung)$tissue <- 'lung'

## Combine lung and blood data and label with batch
gtex_both_lung <- do.call(cbind, list(gtex_blood, gtex_lung))
colData(gtex_both_lung)$batch <- gtex_metadata[
    match(
        colData(gtex_both_lung)$run,
        gtex_metadata$run
    ), 'smgebtch']

## Scale and filter to correct gene set
gtex_both_lung <- scale_counts(gtex_both_lung)
gtex_both_lung_counts <- assays(gtex_both_lung)[[1]]
gtex_both_lung_counts <- gtex_both_lung_counts[keep,]

## Carry out differential expression analysis
design_gtex_lung <- model.matrix(~colData(gtex_both_lung)$tissue +
    colData(gtex_both_lung)$batch)
dge_gtex_lung <- DGEList(counts = gtex_both_lung_counts)
dge_gtex_lung <- calcNormFactors(dge_gtex_lung)
v_gtex_lung <- voom(dge_gtex_lung, design_gtex_lung, plot = FALSE)
fit_gtex_lung <- lmFit(v_gtex_lung, design_gtex_lung)
fit_gtex_lung <- eBayes(fit_gtex_lung)
topTable(fit_gtex_lung, coef = 2)
```


## Compare concordance at the top (CAT) plots

To look for overlap among the most differentially expressed genes across the three analyses, we construct concordance at the top (CAT) plots. We make three comparisons: our five-study results compared to GTEx blood-colon results (same tissue comparison), our five-study results compared to GTEx blood-lung results (different tissue comparison) and our five-study results compared to a ranked set of coefficients for the batch variable in the GTEx blood-lung results (which should not show any biologically meaningful correspondence).


```{r catplots}
## Comparison of initial analysis to GTEx, same tissues
cat_sra_gtex <- CATplot(
    -rank(fit$coefficients[, 2]),
    -rank(-fit_gtex$coefficients[, 2]), maxrank = 1000, ylim = c(0,1), make.plot=FALSE)
## Comparison of initial analysis to GTEx, different tissues
cat_sra_gtex_lung = CATplot(
    -rank(fit$coefficients[, 2]),
    -rank(-fit_gtex_lung$coefficients[, 2]), maxrank = 1000, ylim = c(0,1), make.plot=FALSE)
## Comparison of initial analysis to GTEx, coefficient of batch in model
cat_sra_gtex_batch = CATplot(
    -rank(fit$coefficients[, 2]),
    -rank(-fit_gtex_lung$coefficients[, 3]), maxrank = 1000, ylim = c(0,1), make.plot=FALSE)

plot(cat_sra_gtex, type = 'l', col = trop[1], lwd = 3)
lines(cat_sra_gtex_lung, type = 'l', col = trop[2], lwd = 3)
lines(cat_sra_gtex_batch, type = 'l', col = trop[3], lwd = 3)
legend(0, 0.5, legend=c('Same Tisue', 'Different Tissues', 
    'Tissue vs. Batch'), col = trop[1:3], lwd = 3)
```

From the figure above, we can see that the best concordance is for the comparison to GTEx where we analyzed the same pair of tissues. The GTEx analysis with lung looks considerably less overlapping, with the overlap with the batch coefficient well below the other two.

# Reproducibility

This analysis report was made possible thanks to:

* R `r citep(bib[['R']])`
* `r Biocpkg('BiocStyle')` `r citep(bib[['BiocStyle']])`
* `r Biocpkg('derfinder')` `r citep(bib[['derfinder']])`
* `r CRANpkg('devtools')` `r citep(bib[['devtools']])`
* `r CRANpkg('dplyr')` `r citep(bib[['dplyr']])`
* `r Biocpkg('edgeR')` `r citep(bib[['edgeR']])`
* `r Biocpkg('ffpe')` `r citep(bib[['ffpe']])`
* `r CRANpkg('knitcitations')` `r citep(bib[['knitcitations']])`
* `r CRANpkg('magrittr')` `r citep(bib[['magrittr']])`
* `r Biocpkg('recount')` `r citep(bib[['recount']])`
* `r CRANpkg('rmarkdown')` `r citep(bib[['rmarkdown']])`
* `r Githubpkg('alyssafrazee/RSkittleBrewer')` `r citep(bib[['RSkittleBrewer']])`
* `r Biocpkg('SummarizedExperiment')` `r citep(bib[['SummarizedExperiment']])`
* `r Biocpkg('limma')` `r citep(bib[['voom']])`

[Bibliography file](meta_analysis.bib)

```{r bibliography, results='asis', echo=FALSE, warning = FALSE}
## Print bibliography
bibliography()
```


```{r reproducibility}
## Time spent creating this report:
diff(c(timestart, Sys.time()))

## Date this report was generated
message(Sys.time())

## Reproducibility info
options(width = 120)
devtools::session_info()
```


