---
title: 'Comparison of Recount with GTEx'
author: 'Andrew E. Jaffe and Kasper D. Hansen'
date: 'June 09, 2016'
output:
  BiocStyle::html_document:
    toc: true
  pdf_document:
    toc: true
---

# Overview

This document compares GTEx data release v6 to Recount.  The main issue addressed in this document is mapping up genes and samples between the two datasets.  The annotations are different:

- GTEx uses Gencode v19 mapped to hg19.
- Recount uses UCSC knownGene as represented by the `TxDb.Hsapiens.UCSC.hg38.knownGene` package, mapped to hg38.

# Dependencies

```{r 'setup', echo = FALSE, warning = FALSE, message = FALSE}
timestart <- Sys.time()

## Bib setup
library('knitcitations')

## Load knitcitations with a clean bibliography
cleanbib()
cite_options(hyperlink = 'to.doc', citation_format = 'text', style = 'html')

## Write bibliography information
bibs <- c(
    ballgown = citation('ballgown'),
    BiocStyle = citation('BiocStyle'),
    coop = citation('coop'),
    devtools = citation('devtools'),
    knitcitations = citation('knitcitations'),
    org.Hs.eg.db = citation('org.Hs.eg.db'),
    R = citation(),
    recount = citation('recount'),
    rmarkdown = citation('rmarkdown'),
    stringr = citation('stringr'),
    SummarizedExperiment = citation('SummarizedExperiment')
)

write.bibtex(bibs,
    file = 'compare_with_GTEx.bib')
bib <- read.bibtex('compare_with_GTEx.bib')

## Assign short names
names(bib) <- names(bibs)
```

## R packages

```{r dependencies}
library('SummarizedExperiment')
library('stringr')
library('recount')
library('rtracklayer')
library('org.Hs.eg.db')
library('ballgown')
library('coop')
```

## Data objects

### From Recount

```{r loadRecount}
## Download the GTEx data
if(FALSE ) {
    if(!file.exists(file.path('SRP012682', 'rse_gene.Rdata'))) {
        download_study('SRP012682')
    }
    load(file.path('SRP012682', 'rse_gene.Rdata'))
}
## Temporarely use:
load("/dcl01/leek/data/recount-website/rse/rse_gtex/SRP012682/rse_gene.Rdata")
```



### From GTEx website

We have downloaded the annotation GTF files as well as the raw gene count matrix from the [GTEx portal](http://www.gtexportal.org).

```{r gtexWebsite}
gtexGtf <- import("/dcs01/ajaffe/GTEX/V6/gencode.v19.genes.patched_contigs.gtf")
gtexData <- read_tsv("/dcs01/ajaffe/GTEX/V6/GTEx_Analysis_v6_RNA-seq_RNA-SeQCv1.1.8_gene_reads.gct.gz", skip=2)
gtexCounts <- gtexData[,3:ncol(gtexData)]
rownames(gtexCounts) <- gtexData$Name
```

### From elsewhere

FIXME: We got this from Jaffe, how?

```{r loadObjectsRecount}
load("/dcl01/lieber/ajaffe/PublicData/SRA_GTEX/gtexPd.Rdata")
gtexPd$sra_accession <- str_trim(gtexPd$sra_accession)
gtexPd <- gtexPd[!is.na(gtexPd$sra_accession),]
```

# Mapping GTEx annotation

We map between version by using ENTREZ gene ids.  The Recount representation is already using ENTREZ gene ids, but we need to map GTEx data to ENTREZ.

```{r GTExmap}
gtexMap <- gtexGtf[!duplicated(gtexGtf$gene_id)]
names(gtexMap) <- gtexMap$gene_id
gtexMap <- gtexMap[rownames(gtexCounts)]
stopifnot(all(rownames(gtexMap) == rownames(gtexCounts)))
gtexMap$EnsemblGeneID <- ballgown:::ss(gtexMap$gene_id, "\\.")
eid2ens <- CharacterList(as.list(org.Hs.egENSEMBL2EG))
mm <- match(gtexMap$EnsemblGeneID,names(eid2ens))
gtexMap$EntrezID <- CharacterList(vector("list", length(gtexMap)))
gtexMap$EntrezID[!is.na(mm)] <- eid2ens[mm[!is.na(mm)]]
stopifnot(identical(names(gtexMap), rownames(gtexCounts)))
table(elementLengths(gtexMap$EntrezID))
table(gtexMap$gene_type,elementLengths(gtexMap$EntrezID) == 1)
```

Basically, only protein coding genes have a ENTREZ id.  We keep only the protein coding genes which are uniquely mapped to an ENTREZ id.

```{r keepMapping}
eidIndex <- which(elementLengths(gtexMap$EntrezID)==1 & gtexMap$gene_type == "protein_coding")
gtexMap <- gtexMap[eidIndex,]
gtexCounts <- gtexCounts[eidIndex,]
gtexMap$EntrezID <- sapply(gtexMap$EntrezID,"[", 1)
```

Now, we still have multiple Ensembl gene ids mapping to the same ENTREZ id.  We drop those as well

```{r dropMultiEntrez}
dropIDs <- gtexMap$EntrezID[duplicated(gtexMap$EntrezID)]
keepIdx <- which(!gtexMap$EntrezID %in% dropIDs)
gtexCounts <- gtexCounts[keepIdx,]
gtexMap <- gtexMap[keepIdx,]
rownames(gtexCounts) <- names(gtexMap) <- gtexMap$EntrezID
dim(gtexCounts)
```



# STILL NEEDS TO BE FIXED

```{r}

## keep people w/ counts
fullPd$SAMPID = gsub("-",".", fullPd$SAMPID) 
fullPd = fullPd[match(colnames(gtexCounts), fullPd$SAMPID),]

###########################
## data from recount ######
recountPd <- all_metadata('gtex')
rse_gene2 <- scale_counts(rse_gene)
recountCounts = assays(rse_gene)$counts
recountMap = rowRanges(rse_gene)
recountPd = colData(rse_gene)

## match people
sampleMatch = match(gtexPd$sra_accession,colnames(recountCounts))
gtexPd = gtexPd[!is.na(sampleMatch),]
gtexCounts = gtexCounts[,!is.na(sampleMatch)]
recountCounts = recountCounts[,sampleMatch[!is.na(sampleMatch)]]
identical(gtexPd$sra_accession, colnames(recountCounts))

## match by gene
geneMatch = match(recountMap$gene_id, gtexMap$EntrezID )
recountCounts = recountCounts[!is.na(geneMatch),]
recountMap = recountMap[!is.na(geneMatch)]

gtexMap = gtexMap[geneMatch[!is.na(geneMatch)],]
gtexCounts = gtexCounts[geneMatch[!is.na(geneMatch)],]
identical(rownames(gtexCounts), rownames(recountCounts))
recountPd = recountPd[gtexPd$sra_accession,]

save(gtexCounts, gtexMap, gtexPd, 
     recountPd, recountCounts, recountMap,
     file = "GTEx_websiteAndRecount_EntrezMatched_17559_vs_8551.Rdata",
     compress=TRUE)
```

# Comparison

```{r loadMap}
load("GTEx_websiteAndRecount_EntrezMatched_17559_vs_8551.Rdata")
gtexCounts <- as.matrix(gtexCounts)
```

```{r rawCounts comparisons}
rawCors <- sapply(1:nrow(gtexCounts), function(ii) pcor(gtexCounts[ii,], recountCounts[ii,]))
summary(rawCors)
sum(rawCors<=0.95, na.rm = TRUE)
sum(rawCors<=0.80, na.rm = TRUE)
mean(rawCors >= 0.99, na.rm = TRUE)
```

```{r rawCountsPlot, plot=TRUE}
pdf(file = "rawCorrelations.pdf")
dens <- density(rawCors, from = -1, to = 1, na.rm = TRUE, n = 4096)
plot(dens, xlab = "Pearson correlation", main = "Raw counts")
plot(dens, xlab = "Pearson correlation",
     main = "Raw counts", xlim = c(0.9,1))
```

```{r normCounts}
gtexCounts2 <- log2(sweep(gtexCounts, MARGIN = 2, FUN = "/", colSums(gtexCounts)/(4*10^7)) + 1)
recountCounts2 <- log2(assay(rse_gene2, "counts")[rownames(recountCounts), colnames(recountCounts)] + 1)
normCors <- sapply(1:nrow(gtexCounts2), function(ii) pcor(gtexCounts2[ii,], recountCounts2[ii,]))
summary(normCors)
```

```{r normCountsPlot, plot=TRUE}
pdf(file = "normCors.pdf")
plot(density(normCors, from = -1, to = 1, na.rm = TRUE),
     xlab = "Pearson correlation", main = "Gene counts, normalized for library size")
dev.off()
```

# Differential expression

Between colon and blood

```{r 'inProgress', eval = FALSE}
## This code is incomplete / in progress

gtex_metadata <- all_metadata("gtex")
gtex_metadata2 <- subset(gtex_metadata, smtsd %in% c("Whole Blood", "Colon"))[, c("run", "smtsd", "smgebtch")]
colnames(gtex_metadata2)[c(2,3)] <- c("tissue", "batch")
gtex_metadata2 <- subset(gtex_metadata2, run %in% colnames(recountCounts2))
rownames(gtex_metadata2) <- gtex_metadata2$run
recountCounts2.de <- recountCounts2[, gtex_metadata2$run]
gtexCounts2.de <- gtexCounts2[, gtex_metadata2$run]

## need to get correct counts
design <- model.matrix(~ tissue + batch, levels = gtex_metadata2)


design <- gtex = model.matrix(~colData(gtex <- both)$tissue + colData(gtex <- both)$batch)
dge <- gtex = DGEList(counts=gtex <- both <- counts)
dge <- gtex = calcNormFactors(dge <- gtex)
v <- gtex <- voom(dge <- gtex,design <- gtex,plot=TRUE)
fit <- gtex <- lmFit(v <- gtex,design <- gtex)
fit <- gtex <- eBayes(fit <- gtex)
hist(fit <- gtex$p.value[,2],col=trop[2],main="DE",breaks=100)
```

# Reproducibility

This analysis report was made possible thanks to:

* R `r citep(bib[['R']])`
* `r Biocpkg('ballgown')` `r citep(bib[['ballgown']])`
* `r Biocpkg('BiocStyle')` `r citep(bib[['BiocStyle']])`
* `r CRANpkg('coop')` `r citep(bib[['coop']])`
* `r CRANpkg('devtools')` `r citep(bib[['devtools']])`
* `r CRANpkg('knitcitations')` `r citep(bib[['knitcitations']])`
* `r Biocpkg('org.Hs.eg.db')` `r citep(bib[['org.Hs.eg.db']])`
* `r Githubpkg('leekgroup/recount')` `r citep(bib[['recount']])`
* `r CRANpkg('rmarkdown')` `r citep(bib[['rmarkdown']])`
* `r CRANpkg('stringr')` `r citep(bib[['stringr']])`
* `r Biocpkg('SummarizedExperiment')` `r citep(bib[['SummarizedExperiment']])`

[Bibliography file](compare_with_GTEx.bib)

```{r bibliography, results='asis', echo=FALSE, warning = FALSE}
## Print bibliography
bibliography()
```


```{r reproducibility}
## Time spent creating this report:
diff(c(timestart, Sys.time()))

## Date this report was generated
message(Sys.time())

## Reproducibility info
options(width = 120)
devtools::session_info()
```
