---
title: "recount (gene and exon analyses)"
author: "Kai Kammers and Shannon Ellis"
date: "June 09, 2016"
output: 
  BiocStyle::html_document
  pdf_document:
      toc: true
---


Included is an example of how to download and analyze expression data from SRA study  SRP032798(http://trace.ncbi.nlm.nih.gov/Traces/sra/?study=SRP032798). The data come from human breast cancer samples, and we compare the transcriptomes of triple negative breast cancer(TNBC) and HER2-positive breast cancer samples. Code here demonstrates how to carry out differential gene, exon, expressed region, and exon-exon junction analyses within a single study using 'limma' and 'voom'. We test for concordance among the results of each analysis and demonstrate how to carry out gene ontology analysis using 'topGO' to characterize top hits from differential expression anlyses

We first load the required packages.

```{r load-packages, message=FALSE}
## load libraries
library(recount)
library(SummarizedExperiment)
library(limma)
library(edgeR)
library(qvalue)
library(topGO)
library(matrixStats)
library(RSkittleBrewer)
library(derfinder)
```



# Gene level analysis

We first download the project of interest (SRP032798), obtaining expression data for the study of interest. We obtain summaries of the number of samples and genes included using colData() and rowData(), respectively.

```{r download-genes}
## Find the project of interest (SRP032798), e.g. with parts of the abstract
project_info <- abstract_search('To define the digital transcriptome of three breast cancer')

## Explore information
project_info

## Browse the project at SRA
browse_study(project_info$project)

## Download the gene-level RangedSummarizedExperiment data
download_study(project_info$project)

## Load the data
load(file.path(project_info$project, 'rse_gene.Rdata'))
rse_gene

## This is the phenotype data provided by the recount project
colData(rse_gene)

## At the gene level, the row data includes the names of the genes and
## the sum of the reduced exons widths, which can be used for taking into
## account the gene length.
rowData(rse_gene)
```

Downloaded count data are first scaled to take into account differing coverage between samples. Phenotype data (pheno) are obtained and ordered to match the sample order of the gene expression data (rse_gene). Only those samples that are HER2-positive or TNBC are included for analysis. Prior to differential gene expression analysis, count data are obtained in matrix format and then filtered to only include those genes with greater than five average normalized counts across all samples.

```{r filter-genes}
## Scale counts by taking into account the total coverage per sample
rse <- scale_counts(rse_gene)

## download additional phenotype data from 
## http://trace.ncbi.nlm.nih.gov/Traces/study/?acc=SRP032789
pheno <- read.table('SraRunTable_SRP032789.txt', sep='\t', 
                    header=TRUE,
                    stringsAsFactors = FALSE)

## obtain correct order for pheno data
pheno <- pheno[match(rse$run, pheno$Run_s), ]
identical(pheno$Run_s, rse$run)
head(cbind(pheno$Run_s, rse$run))

## obtain grouping information
colData(rse)$group <- pheno$tumor_type_s
table(colData(rse)$group)   
    
## subset data to HER2 and TNBC types
rse <- rse[, rse$group %in% c("HER2 Positive Breast Tumor", "TNBC Breast Tumor")]
rse 

## obtain count matrix
counts <- assays(rse)$counts

## filter count matrix
filter <- apply(counts, 1, function(x) mean(x) > 5)
counts <- counts[filter, ]
dim(counts)
counts_genes <- counts
```

To get a better sense of the data, we plot the mean-variance relationship for each gene. Similarly, we run principal component analysis (PCA) to identify any sample outliers within the data. We assess the variance explained by each of the first 11 PCs as well as visualize the relationship of each sample in the first two PCs.

```{r exploratory-genes, fig.height=8, fig.width=8}
## set colors 
trop <- RSkittleBrewer('tropical')[c(1,2)]
cols <- as.numeric(as.factor(rse$group))

## Look at mean variance relationship
plot(rowMeans(log2(counts+1)), rowVars(log2(counts+1)),
     pch=19, col=trop[2])

## calculate PCs with svd function
expr.pca <- svd(counts - rowMeans(counts))

## plot variance explained by PCs
par(font.lab=2, cex.lab=1.2, font.axis=2, cex.axis=1.2)
plot(expr.pca$d^2/sum(expr.pca$d^2), pch=19, col=trop[2], cex=1.5,
     ylab="Fraction of variance explained (gene level)", xlab="PC #",
     main="PCs")

##plot PC1 vs. PC2
par(font.lab=2, cex.lab=1.2, font.axis=2, cex.axis=1.2)
plot(expr.pca$v[, 1], expr.pca$v[, 2], pch=19, col=trop[cols], cex=1.5,
     xlab="PC1", ylab="PC2",
     main="PC (gene level)")
legend("topright", pch=19, col=trop[c(1,2)],
       names(summary(as.factor(rse$group))))
```

Having determined there are no sample outliers in these data, we carry out differential gene expression analysis. Differential gene expression between TNBC and HER2-positive samples are determined using limma and voom. Differentially expressed genes are visualized using a volcano plot to compare the effect size of the differential expression (as measured by the log2(fold change) in expression) and its significance (-log10p-value).

```{r voom-genes, fig.height=8, fig.width=8}
## Perform differential expression analysis with limma-voom
design <- model.matrix(~ rse$group)
design

dge <- DGEList(counts=counts)
dge <- calcNormFactors(dge)
v <- voom(dge,design,plot=TRUE)
fit <- lmFit(v,design)
fit <- eBayes(fit)
log2FC <- fit$coefficients[, 2]
p.mod <- fit$p.value[, 2]
q.mod <- qvalue(p.mod)$q
res.genes <- data.frame(log2FC, p.mod, q.mod)
rownames(res.genes) <- rownames(counts)
#str(as.character(rse@rowRanges@seqnames))
#determine the number of genes differentially expressed q<0.05
sum(res.genes$q.mod<0.05)


## Histogram
par(font.lab=2, cex.lab=1.2, font.axis=2, cex.axis=1.2)
hist(p.mod, col=trop[2], xlab="p-value",
     main="Histogramm of p-values", breaks=100)


## Volcano plot
par(font.lab=2, cex.lab=1.2, font.axis=2, cex.axis=1.2)
rx2 <- c(-1, 1)*1.1*max(abs(log2FC))
ry2 <- c(-0.1, max(-log10(p.mod)))*1.1
plot(log2FC, -log10(p.mod), 
     pch=19, xlim=rx2, ylim=ry2, col=trop[2],
     xlab=bquote(paste(log[2], " (fold change)")), 
     ylab=bquote(paste(-log[10], " (p-value)")))
abline(v=seq(-10,10,1), col="lightgray", lty="dotted")
abline(h=seq(0,23,1), col="lightgray", lty="dotted")
points(log2FC, -log10(p.mod), pch=19, col=trop[2])
title("Volcano plot: TNBC vs. HER2+ in SRP032798 (gene level)")
```



# Gene set analysis
To get a better understanding of those genes showing differential gene expression, we utilize `topGO`, a gene set analysis library. Genes included in this analysis are those reaching a q-value cutoff less than 0.05. 

```{r topgo-prep} 
names(q.mod) <- rownames(counts)
interesting <- function(x) x < 0.05
```

After determining which genes to include for analysis, `topGO` objects are generated and the enrichment tests are run. The Kolomogorov-Smirnov (`ks`) test is used to test for distributional differences. Here, we ask whether each GO group is "enriched" for differentially expressed (q.mod < 0.05) genes. Equivalently, we are testing whether the p-value distributions are the same for genes in and outside of each gene ontology. We run tests on the "biological processes" ontology. 

```{r bp}
topgoobjBP <- new('topGOdata', 
    description='biological process',
    ontology='BP', allGenes=q.mod, geneSelectionFun=interesting, 
    annotationFun=annFUN.org, mapping='org.Hs.eg.db', ID='entrez')

bptest <- runTest(topgoobjBP, algorithm='weight01', statistic='ks')
bptest
bpres_gene <- GenTable(topgoobjBP, pval=bptest, topNodes=length(bptest@score), numChar=100)
head(bpres_gene, n=10)                  
```



# Exon level analysis
**As above, we're interested here in differential expression. However, rather than summarizing across genes, this analysis will look for differential expression at the exon level. In this analysis, we include all exons that map to the previous filtered genes and again carry out differential expression analysis  using limma and voom.**

Here, we download data from the same project as above (SRP032798); however, this time, we are interested in obtaining the exon-level data.

```{r download-exons}
## Find a project of interest (SRP032798)
project_info <- abstract_search('To define the digital transcriptome of three breast cancer')
project_info

## Browse the project at SRA
browse_study(project_info$project)

## Download the exon-level RangedSummarizedExperiment data
download_study(project_info$project, type="rse-exon")

## Load the data
load(file.path(project_info$project, 'rse_exon.Rdata'))
rse_exon

## This is the sample phenotype data provided by the recount project
colData(rse_exon)
```

As above, downloaded count data are first scaled to take into account differing coverage between samples. The same phenotype data (pheno) are used and again ordered to match the sample order of the expression data (rse_exon). Only those samples that are HER2-positive or TNBC are included for analysis. Prior to differential exon expression analysis, count data are obtained in matrix format and then filtered to only include exons within genes that had been analyzed previously.


```{r exons}
## Scale counts by taking into account the total coverage per sample
rse <- scale_counts(rse_exon)

## download pheno data from 
## http://trace.ncbi.nlm.nih.gov/Traces/study/?acc=SRP032798
pheno <- read.table('SraRunTable_SRP032789.txt', sep='\t', 
                    header=TRUE,
                    stringsAsFactors = FALSE)

## obtain correct order for pheno data
pheno <- pheno[match(rse$run, pheno$Run_s), ]
identical(pheno$Run_s, rse$run)
head(cbind(pheno$Run_s, rse$run))

## obtain grouping information
colData(rse)$group <- pheno$tumor_type_s
table(colData(rse)$group)   
    
## subset data to HER2 and TNBC types
rse <- rse[, rse$group %in% c("HER2 Positive Breast Tumor", "TNBC Breast Tumor")]
rse 
rse.exon.filt <- rse

## obtain count matrix
counts <- assays(rse)$counts
dim(counts)


## filter count matrix (keep exons that are in filtered gene counts matrix)
filter <- rownames(counts) %in% rownames(counts_genes)
counts <- counts[filter, ]
dim(counts)
```
As above, to get a better sense of the data, we assess the mean-variance relationship for each exon. Similarly, we run principal component analysis (PCA) to identify any sample outliers within the data. We assess the variance explained by each of the first 11 PCs as well as visualize the relationship of each sample in the first two PCs.


```{r exploratory-exons, fig.height=8, fig.width=8}
## set colors 
trop <- RSkittleBrewer('tropical')[c(1,2)]
cols <- as.numeric(as.factor(rse$group))

## Look at mean variance relationship
plot(rowMeans(log2(counts+1)), rowVars(log2(counts+1)),
     pch=19, col=trop[2])

## calculate PCs with svd function
expr.pca <- svd(counts - rowMeans(counts))

## plot PCs
par(font.lab=2, cex.lab=1.2, font.axis=2, cex.axis=1.2)
plot(expr.pca$d^2/sum(expr.pca$d^2), pch=19, col=trop[2], cex=1.5,
     ylab="Fraction of variance explained", xlab="PC #",
     main="PCs (exon level)")

##plot PC1 vs. PC2
par(font.lab=2, cex.lab=1.2, font.axis=2, cex.axis=1.2)
plot(expr.pca$v[, 1], expr.pca$v[, 2], pch=19, col=trop[cols], cex=1.5,
     xlab="PC1", ylab="PC2",
     main="PC (exon level)")
legend("topright", pch=19, col=trop[c(1,2)],
       names(summary(as.factor(rse$group))))
```

Again, differential expression analysis is carried out using limma and voom; however, this time at the exon, rather than gene, level. Data are again visualized using a volcano plot to assess the strength (log2FC) and statistical significance (-log10p-value) for each exon.

```{r voom-exons, fig.height=8, fig.width=8}
design <- model.matrix(~ rse$group)
design

dge <- DGEList(counts=counts)
dge <- calcNormFactors(dge)
v <- voom(dge,design,plot=TRUE)
fit <- lmFit(v,design)
fit <- eBayes(fit)
log2FC <- fit$coefficients[, 2]
p.mod <- fit$p.value[, 2]
q.mod <- qvalue(p.mod)$q
res.exons <- data.frame(log2FC, p.mod, q.mod)
#determine the number of exons differentially expressed q<0.05
sum(res.exons$q.mod<0.05)

## Volcano plot
par(font.lab=2, cex.lab=1.2, font.axis=2, cex.axis=1.2)
rx2 <- c(-1, 1)*1.1*max(abs(log2FC))
ry2 <- c(-0.1, max(-log10(p.mod)))*1.1
plot(log2FC, -log10(p.mod), 
     pch=19, xlim=rx2, ylim=ry2, col=trop[2],
     xlab=bquote(paste(log[2], " (fold change)")), 
     ylab=bquote(paste(-log[10], " (p-value)")))
abline(v=seq(-10,10,1), col="lightgray", lty="dotted")
abline(h=seq(0,23,1), col="lightgray", lty="dotted")
points(log2FC, -log10(p.mod), pch=19, col=trop[2])
title("Volcano plot: TNBC vs. HER2+ in SRP032798 (exon level)")
```

To compare findings at the gene and exon level, we obtained a single exon-level p-value for each gene included at the gene-level analysis. To do this, we utilized Simes' rule [PMID: 23330866], such that for each gene included in the gene-level analysis, the p-values for exons within that gene were extracted and sorted. Each exon-level p-value is then multiplied by the number of exons present within the gene. For each exon (1,2...n), this quantity is divided by that exon's rank (where 1=most significant exon and n=least significant). The minimum value from this calcultion is assigned as the exon-level p-value at each gene.

```{r, exons-to-genes}
gene_id <- unique(rownames(counts))

## calculate p-values for gens with Simes' rule
p_gene <- NULL
for(i in 1:length(gene_id)){
        gene_id[i]
    p_exon <- res.exons$p.mod[rownames(counts) %in% gene_id[i]]
    p_exon <- sort(p_exon)
    p_exon_simes <- NULL 
    for(j in 1:length(p_exon)){
        p_exon_simes[j] <- length(p_exon) * p_exon[j] / j
    }
    p_gene[i] <- min(p_exon_simes)
}

names(p_gene) <- gene_id
q_gene <- qvalue(p_gene)$q
#determine the number of gene-level exons differentially expressed q<0.05
sum(q_gene<0.05)

As above, 'topGO' can be utilized to assign biological function to differentially expressed exons.

## Gene set analysis (p-values of genes derived with Simes' rule from exon p-values)
interesting <- function(x) x < 0.05

topgoobjBP <- new('topGOdata', 
    description='biological process',
    ontology='BP', allGenes=q_gene, geneSelectionFun=interesting, 
    annotationFun=annFUN.org, mapping='org.Hs.eg.db', ID='entrez')

bptest <- runTest(topgoobjBP, algorithm='weight01', statistic='ks')
bptest
bpres_exon <- GenTable(topgoobjBP, pval=bptest, topNodes=length(bptest@score), numChar=100)
head(bpres_exon, n=10)                  
```

To determine the concordance between the gene- and exon-level analyses, the top hits (as determined by p-value) are compared. Results are plotted such that the points falling along the identity line would indicate complete agreement between the top hits of each analysis. 

# Concordance (genes)
```{r, con, fig.width=8, fig.height=8}
## obtain and sort p-values for genes
p.mod1 <- res.genes$p.mod
names(p.mod1) <- rownames(res.genes)
p.mod1.sort <- p.mod1[order(p.mod1)]


## obtain dnd sort p-values for genes derived from exons
p.mod2 <- p_gene
p.mod2.sort <- p.mod2[order(p.mod2)]


## overlap for genes between studies
table(names(p.mod1.sort) %in% names(p.mod2.sort))
table(names(p.mod2.sort) %in% names(p.mod1.sort))

conc <- NULL
for(i in 1:length(p.mod1.sort)){
    conc[i] <- sum(names(p.mod1.sort)[1:i] %in% names(p.mod2.sort)[1:i])
}


par(mfrow=c(1,1), font.lab=1.5, cex.lab=1.2, font.axis=1.5, cex.axis=1.2)
plot(seq(1:length(p.mod1.sort)), conc, 
     type="l", las=0,
     xlim=c(0,18000),
     ylim=c(0,18000),
     xlab='ordered genes',
     ylab='ordered genes (from exon analysis)',
     main='Concordance')
for(k in 1:3){
    abline(v=k*5000, cex=0.5, col="lightgrey")
    abline(h=k*5000, cex=0.5, col="lightgrey")
}
lines(seq(1:length(p.mod1.sort)), conc, col="black", lwd=2)


par(mfrow=c(1,1), font.lab=1.5, cex.lab=1.2, font.axis=1.5, cex.axis=1.2)
plot(seq(1:length(p.mod1.sort[1:100])), conc[1:100], 
     type="l",
     xlim=c(0,100),
     ylim=c(0,100),
     xlab='ordered genes',
     ylab='ordered genes (from exon analysis)',
     main='Concordance')
for(k in 1:5){
    abline(v=k*20, cex=0.5, col="lightgrey")
    abline(h=k*20, cex=0.5, col="lightgrey")
}
lines(seq(1:length(p.mod1.sort[1:100])), conc[1:100], col="black", lwd=2)
```

Concordance can also be calculated looking at the gene ontology (GO) groups identified from the gene and exon-level analyses. Again, we plot the agreement between the two analyses such that complete agreement between the two analyses would fall along the identity line.

# Concordance (GO groups)
```{r, con-GO, fig.width=8, fig.height=8}
## obtain and sort p-values for genes
p.mod1 <- bpres_gene$GO.ID
names(p.mod1) <- bpres_gene$GO.ID
p.mod1.sort <- p.mod1


## obtain and sort p-values for genes derived from exons
p.mod2 <- bpres_exon$GO.ID
names(p.mod2) <- bpres_exon$GO.ID
p.mod2.sort <- p.mod2


## overlap for genes between studies
table(names(p.mod1.sort) %in% names(p.mod2.sort))
table(names(p.mod2.sort) %in% names(p.mod1.sort))

conc <- NULL
for(i in 1:length(p.mod1.sort)){
    conc[i] <- sum(names(p.mod1.sort)[1:i] %in% names(p.mod2.sort)[1:i])
}


par(mfrow=c(1,1), font.lab=1.5, cex.lab=1.2, font.axis=1.5, cex.axis=1.2)
plot(seq(1:length(p.mod1.sort)), conc, 
     type="l", las=0,
     xlim=c(0,18000),
     ylim=c(0,18000),
     xlab='ordered GO groups',
     ylab='ordered GO groups (from exon analysis)',
     main='Concordance')
for(k in 1:3){
    abline(v=k*5000, cex=0.5, col="lightgrey")
    abline(h=k*5000, cex=0.5, col="lightgrey")
}
lines(seq(1:length(p.mod1.sort)), conc, col="black", lwd=2)


par(mfrow=c(1,1), font.lab=1.5, cex.lab=1.2, font.axis=1.5, cex.axis=1.2)
plot(seq(1:length(p.mod1.sort[1:100])), conc[1:100], 
     type="l",
     xlim=c(0,100),
     ylim=c(0,100),
     xlab='ordered GO groups',
     ylab='ordered GO groups (from exon analysis)',
     main='Concordance')
for(k in 1:5){
    abline(v=k*20, cex=0.5, col="lightgrey")
    abline(h=k*20, cex=0.5, col="lightgrey")
}
lines(seq(1:length(p.mod1.sort[1:100])), conc[1:100], col="black", lwd=2)
```


# Technical reproducibility information
```{r TRI}
## Reproducibility info
proc.time()
message(Sys.time())
options(width = 120)
devtools::session_info()
```

This document was processed on: `r Sys.Date()`.
