---
title: 'recount (DER analyses)'
author: 'Kai Kammers and Shannon Ellis'
date: 'July 22, 2016'
output:
  BiocStyle::html_document:
    toc: true
  pdf_document:
    toc: true
---

```{r 'setup', echo = FALSE, warning = FALSE, message = FALSE}
timestart <- Sys.time()

## Bib setup
library('knitcitations')

## Load knitcitations with a clean bibliography
cleanbib()
cite_options(hyperlink = 'to.doc', citation_format = 'text', style = 'html')

## Write bibliography information
bibs <- c(
    BiocParallel = citation('BiocParallel'),
    BiocStyle = citation('BiocStyle'),
    derfinder = citation('derfinder')[1],
    devtools = citation('devtools'),
    downloader = citation('downloader'),
    edgeR = citation('edgeR')[1],
    GenomicRanges = citation('GenomicRanges'),
    knitcitations = citation('knitcitations'),
    matrixStats = citation('matrixStats'),
    R = citation(),
    recount = citation('recount'),
    rmarkdown = citation('rmarkdown'),
    RSkittleBrewer = citation('RSkittleBrewer'),
    SummarizedExperiment = citation('SummarizedExperiment'),
    topGO = citation('topGO'),
    voom = RefManageR::BibEntry('article', key = 'voom', author = 'CW Law and Y Chen and W Shi and GK Smyth', year = '2014', title = 'Voom: precision weights unlock linear model analysis tools for RNA-seq read counts', journal = 'Genome Biology', volume = '15', pages = 'R29')
)

write.bibtex(bibs,
    file = 'recount_DER_SRP032789.bib')
bib <- read.bibtex('recount_DER_SRP032789.bib')

## Assign short names
names(bib) <- names(bibs)
```

Here is an example of how to download and analyze a `RangedSummarizedExperiment` object. While differential expression analysis of RNA-Seq data is often done at the gene and/or exon level, there is utility in looking for differential expression using an annotation-indepenent approach. Here, without the aid of any annotation, expressed regions are summarized, filtered, and differential expression analyses are carried out. These results are compared to those from annotation-dependent analyses carried out on the same data to assess whether additional information can be garnered from this approach.
Data herein come from triple negative and HER2-positive breast cancer samples from SRA study id [SRP032798](http://trace.ncbi.nlm.nih.gov/Traces/sra/?study=SRP032798). 

We first load the required packages.

```{r load-packages, message = FALSE, warning = FALSE}
## load libraries
library('recount')
library('SummarizedExperiment')
library('limma')
library('edgeR')
library('qvalue')
library('topGO')
library('matrixStats')
library('RSkittleBrewer')
library('derfinder')
library('derfinderPlot')
library('BiocParallel')
library('GenomicRanges')
library('bumphunter')
library('downloader')
library('TxDb.Hsapiens.UCSC.hg38.knownGene')

## set colors 
trop <- RSkittleBrewer('tropical')[c(1, 2)]
```

# Download study data
Expressed region data are downloaded for each chromosome and summarized into a count matrix (`covMat`) where each region is a row and each sample a column.

```{r download-study-data}

chrs <- paste0('chr', c(1:22, 'X', 'Y'))
bp <- SerialParam() ## Change if you have access to more cores

## Download region level data
if(!file.exists('regions_SRP032789.Rdata')) {
    regions_list <- bplapply(chrs, function(chr) {
        regs <- expressed_regions('SRP032789', chr, cutoff = 5L,
            maxClusterGap = 3000L, verbose = FALSE)
        return(regs)
    }, BPPARAM = bp)
    names(regions_list) <- chrs
    regions <- unlist(GRangesList(regions_list))
    
    ## Save the regions
    save(regions, regions_list, file = 'regions_SRP032789.Rdata')
} else {
    load('regions_SRP032789.Rdata')
}

## Compute coverage matrix for study SRP032789
if(!file.exists('covMat_SRP032789.Rdata')) {
    covMat <- bplapply(chrs, function(chr) {
        coverageMatrix <- coverage_matrix('SRP032789', chr,
            regions_list[[chr]], verboseLoad = FALSE)
        return(coverageMatrix)
    }, BPPARAM = bp)
    covMat <- do.call(rbind, covMat)

    ## Round the coverage matrix to integers
    covMat <- round(covMat, 0)
    save(covMat, file = 'covMat_SRP032789.Rdata')
} else {
    load('covMat_SRP032789.Rdata')
}

```

# Obtain count data 
After summarizing data across expressed regions, data are filtered to only include samples of interest (TNBC and HER2+ tumor samples, (`covMat_filt`) and regions with greater than 5 mean read counts across samples (`counts`). 


```{r filter-er}
## Download phenotype data from 
## http://trace.ncbi.nlm.nih.gov/Traces/study/?acc=SRP032789
pheno <- read.table('SraRunTable_SRP032789.txt', sep = '\t', 
    header=TRUE,
    stringsAsFactors = FALSE)

## Check ordering of samples
pheno <- pheno[pheno$Run_s %in% colnames(covMat), ]
identical(pheno$Run_s, colnames(covMat))

## Obtain correct order for pheno data
pheno <- pheno[match(colnames(covMat), pheno$Run_s), ]
identical(pheno$Run_s, colnames(covMat))
head(cbind(pheno$Run_s, colnames(covMat)))

## Find tumor type information
group <- pheno$tumor_type_s
table(group)   
    
## Subset data to HER2 and TNBC type
covMat_filt <- covMat[, group %in% c('HER2 Positive Breast Tumor', 'TNBC Breast Tumor')]
group <- group[group %in% c('HER2 Positive Breast Tumor', 'TNBC Breast Tumor')] 
dim(covMat_filt)
rownames(covMat_filt) <- rownames(regions)

## Filter count matrix
counts <- covMat_filt
filter <- apply(counts, 1, function(x) mean(x) > 5)
counts <- counts[filter, ]
dim(counts)

## Obtain chromosome and position information for regions included after filtering
regions_counts <- regions[filter, ]
```

# DE analysis
Differential expression analysis is carried out at the expressed-region level using `voom`
and `limma`. Results for each region are plotted using a volcano plot to compare the effect size of the differential expression [ as measured by the $log_2(fold-change)$ in expression ] and its significance [ $-log_10(p-value)$ ].


```{r voom-er, fig.height = 8, fig.width = 8}
design <- model.matrix(~ group)
design

dge <- DGEList(counts = counts)
dge <- calcNormFactors(dge)
v <- voom(dge, design, plot = TRUE)
fit <- lmFit(v, design)
fit <- eBayes(fit)
log2FC <- fit$coefficients[, 2]
p.mod <- fit$p.value[, 2]
q.mod <- qvalue(p.mod)$q
res.regions <- data.frame(log2FC, p.mod, q.mod)
sum(res.regions$q.mod < 0.05)

## Add differential expression information to GRanges object containing information 
## about where the expressed regions are located in the genome (regions_counts)
regions_counts$log2FC <- res.regions$log2FC
regions_counts$p.mod <- res.regions$p.mod
regions_counts$q.mod <- res.regions$q.mod

## Volcano plot
par(font.lab = 2, cex.lab = 1.2, font.axis = 2, cex.axis = 1.2)
rx2 <- c(-1, 1) * 1.1 * max(abs(log2FC))
ry2 <- c(-0.1, max(-log10(p.mod))) * 1.1
plot(log2FC, -log10(p.mod), 
     pch = 19, xlim = rx2, ylim = ry2, col = trop[2],
     xlab = bquote(paste(log[2], ' (fold change)')), 
     ylab = bquote(paste(-log[10], ' (p-value)')))
abline(v = seq(-10, 10, 1), col = 'lightgray', lty = 'dotted')
abline(h = seq(0, 23, 1), col = 'lightgray', lty = 'dotted')
points(log2FC, -log10(p.mod), pch = 19, col = trop[2])
title('Volcano plot: TNBC vs. HER2+ in SRP032789 (er level)')
```


# Annotate regions with genomic information
After detecting regions differentially expressed between TNBC and HER2+ samples, we can annotate where in the genome these regions are found and compare them back to the results from the gene- and exon-level differential expression analyses to determine what new information is gained by utilziing expressed region level data.


```{r annotation-er}
## Obtain significant (q<0.05) regions
regions_counts_sig <- regions_counts[regions_counts$q.mod < 0.05, ]

## Identify the top regions by highest total coverage
# top <- regions_counts_sig[order(regions_counts_sig$area, decreasing = TRUE)[1:100], ]

## Annotate significant regions
genes <- annotateTranscripts(TxDb.Hsapiens.UCSC.hg38.knownGene)

## Match detected regions with known gene annotation
## Warning: This can take some time
## (dependent upon how many regions are included for annotation)
# annotation_der <- matchGenes(regions_counts_sig, genes)

## Merge DER information with annotation information
# regions_annotated <- cbind(regions_counts_sig, annotation_der)
```

There are `r length(regions_counts_sig)` statistically significant regions detected (`q < 0.05`). We can use a venn diagram to visually represent where they are found in the genome.

```{r visualizing-DERs, fig.height = 8, fig.width = 8}

## Get Annotation
if(!file.exists('genomicState.Hsapiens.BioMart.ENSEMBLMARTENSEMBL.GRCh38.p5.Rdata')) {
## Genomic state created by https://github.com/nellore/runs/blob/master/gtex/DER_analysis/coverageMatrix/genomicState/hg38-genomicState.R
    download('https://github.com/nellore/runs/blob/master/gtex/DER_analysis/coverageMatrix/genomicState/genomicState.Hsapiens.BioMart.ENSEMBLMARTENSEMBL.GRCh38.p5.Rdata?raw=true', mode = 'wb', destfile = 'genomicState.Hsapiens.BioMart.ENSEMBLMARTENSEMBL.GRCh38.p5.Rdata')
}

load('genomicState.Hsapiens.BioMart.ENSEMBLMARTENSEMBL.GRCh38.p5.Rdata')
gs_raw <- genomicState.Hsapiens.BioMart.ENSEMBLMARTENSEMBL.GRCh38.p5$fullGenome
gs <- renameSeqlevels(gs_raw, paste0('chr', seqlevels(gs_raw)))
gs_exons <- gs[gs$theRegion =="exon"]

## Make venn diagram
## Venn Diagram for significant DERs (q<0.05)
annoRegs_sigDER <- annotateRegions(regions_counts_sig, gs, minoverlap = 20)
#pdf(file = 'plots/VennDiagram_sigDERs.pdf')
vennRegions(annoRegs_sigDER, main = 'Significant DERs (q<0.05)', counts.col = trop[1])
#dev.off()
```

# Gene level analysis
In order to compare these differentially expressed regions with gene-level results, we must carry out differential expression in these same data at the gene level.

We first download the project of interest (`SRP032798`), obtaining expression data for the study of interest.

```{r download-genes}
## Find the project of interest (SRP032789), e.g. with parts of the abstract
project_info <- abstract_search('To define the digital transcriptome of three breast cancer')

## Download the gene-level RangedSummarizedExperiment data
if(!file.exists(file.path('SRP032789', 'rse_gene.Rdata'))) {
    download_study(project_info$project)
}

## Load the data
load(file.path(project_info$project, 'rse_gene.Rdata'))
rse_gene
```

# QC data
Downloaded count data are first scaled to take into account differing coverage between samples. Phenotype data (`pheno`) are obtained and ordered to match the sample order of the gene expression data (`rse_gene`). Only those samples that are HER2-positive or TNBC are included for analysis. Prior to differential gene expression analysis, count data are obtained in matrix format and then filtered to only include those genes with greater than five average normalized counts across all samples.

```{r filter-genes}
## Scale counts by taking into account the total coverage per sample
rse <- scale_counts(rse_gene)

## Download additional phenotype data from 
## http://trace.ncbi.nlm.nih.gov/Traces/study/?acc=SRP032789
pheno <- read.table('SraRunTable_SRP032789.txt', sep = '\t', 
    header = TRUE,
    stringsAsFactors = FALSE)

## Obtain correct order for pheno data
pheno <- pheno[match(rse$run, pheno$Run_s), ]
identical(pheno$Run_s, rse$run)
head(cbind(pheno$Run_s, rse$run))

## Obtain grouping information
colData(rse)$group <- pheno$tumor_type_s
table(colData(rse)$group)   
    
## Subset data to HER2 and TNBC types
rse <- rse[, rse$group %in% c('HER2 Positive Breast Tumor',
    'TNBC Breast Tumor')]
rse 

## Obtain count matrix
counts <- assays(rse)$counts

## Filter count matrix
filter <- apply(counts, 1, function(x) mean(x) > 5)
counts <- counts[filter, ]
dim(counts)
counts_genes <- counts
```

# DE analysis : gene-level
Using `limma` and `voom`, differentially expressed genes are detected at the gene level.

```{r voom-genes, fig.height = 8, fig.width = 8}
## Perform differential expression analysis with limma-voom
design <- model.matrix(~ rse$group)
design

dge <- DGEList(counts = counts)
dge <- calcNormFactors(dge)
v <- voom(dge, design, plot = TRUE)
fit <- lmFit(v, design)
fit <- eBayes(fit)
log2FC <- fit$coefficients[, 2]
p.mod <- fit$p.value[, 2]
q.mod <- qvalue(p.mod)$q
res.genes <- data.frame(log2FC, p.mod, q.mod)
rownames(res.genes) <- rownames(counts)
sum(res.genes$q.mod < 0.05)
```

# Compare DER findings to gene-level analysis
By avoiding using gene or exon-level annotation for quantification, regions usually excluded from analysis can be included. Here, the same data set is utilized for differential expression; however, we determine the number of signficantly differentially expressed regions (DERs, q<0.05) that do not overlap any gene. These regions are places within the genome that would be completely missed by the common gene-level analyses. 

```{r DERs-undetected, fig.height = 8, fig.width = 8}
## Determine DERs not detected by gene level analysis
## Get DEGs into GRanges object format
genes_GRanges <- SummarizedExperiment::rowRanges(rse_gene[filter, ])
genes_GRanges$log2FC <- res.genes$log2FC
genes_GRanges$p.mod <- res.genes$p.mod
genes_GRanges$q.mod <- res.genes$q.mod

## For each region, determine if region overlaps with a gene included in the gene-level analysis
overlap <-countOverlaps(regions_counts_sig,genes_GRanges,type="any")
overlap <-countOverlaps(regions_counts,genes_GRanges,type="any")

regions_counts$overlap <- overlap
regions_counts_ingene <- regions_counts[overlap>0]
regions_counts_notingene <- regions_counts[overlap==0]

## For significant DERs (q<0.05), how many sites do not fall within a gene 
## (and thus would not be detected by gene-level analyses
 
## number of DERs that do not overlap with genes in gene-level analysis
length(regions_counts_notingene[regions_counts_notingene$q.mod<0.05])

## Volcano plot of DERs (that do not overlap with genes)
par(font.lab = 2, cex.lab = 1.2, font.axis = 2, cex.axis = 1.2)
rx2 <- c(-11, 11)
ry2 <- c(-0.1, max(-log10(p.mod))) * 1.1
plot(log2FC, -log10(p.mod), 
     pch = 19, xlim = rx2, ylim = ry2, col = trop[2],
     xlab = bquote(paste(log[2], ' (fold change)')), 
     ylab = bquote(paste(-log[10], ' (p-value)')))
abline(v = seq(-10, 10, 1), col = 'lightgray', lty = 'dotted')
abline(h = seq(0, 23, 1), col = 'lightgray', lty = 'dotted')
points(regions_counts_notingene$log2FC, -log10(regions_counts_notingene$p.mod), pch = 19, col = trop[2])
title('Volcano plot: TNBC vs. HER2+ in SRP032789 (DERs not in genes)')
```

There are `r length(regions_counts_notingene[regions_counts_notingene$q.mod<0.05])` DERs that do not overlap with genes included in the gene-level analysis.

These range in size from `r min(width(regions_counts_notingene))` bps to `r max(width(regions_counts_notingene))` bps in length, with a median length of `r median(width(regions_counts_notingene))` bps.

# Compare DERs to known genomic annotation
Beyond an interest in those regions not detected in the gene-level analysis, we are also interested in those regions that do not overlap any annotated gene. First, we quantify where DERs are found within the genome that do not overlap any known exons. Then, we highlight single locus coverage plots for a few of the longest DERs.


```{r visualizing-DERs-that-do-not-overlap-annotated-exons, fig.height = 11, fig.width = 11}

## For each region, determine if region overlaps with a gene included
overlapANY <-countOverlaps(regions_counts,gs_exons,type="any")

regions_counts$overlapANY <- overlapANY
regions_counts_ingeneANY <- regions_counts[regions_counts$overlapANY>0]
regions_counts_notinANYgene <- regions_counts[regions_counts$overlapANY==0]

## Number of DERs that do not overlap with any genes
length(regions_counts_notinANYgene[regions_counts_notinANYgene$q.mod<0.05])


## Get top 1% longest regions outside of genes to visualize
regions = subset(regions_counts_notinANYgene, width(regions_counts_notinANYgene) > quantile(width(regions_counts_notinANYgene),0.99))
#order so most significant are plotted first
regions <- regions[order(regions$p.mod), ]
    
## make region plots
tIndexes = split(1:length(rse$group), rse$group)

## Get required information for the plots
## Annotate regions
annoRegs <- annotateRegions(regions, gs, minoverlap = 1)

## Load full coverage
SRP032789_urls <- subset(recount_url, project == 'SRP032789' & file_name != 'mean_SRP032789.bw')
#### Leo! change recount_url$path to recount_url$url for submission ####
files <- SRP032789_urls$url[grepl('bw$', SRP032789_urls$url)]
names(files) <- gsub('.bw', '', SRP032789_urls$file_name[grepl('bw$', SRP032789_urls$url)])

## Only include TNBC and HER2+ samples
files <- subset(files, names(files) %in% colnames(covMat_filt))

## Obtain fullCoverage matrix for all chromosomes
fullCov <- fullCoverage(files = files, chrs = paste0('chr', c(1:22, 'X', 'Y')),
    which = regions, verbose = FALSE)

## Find nearest annotation with bumphunter::matchGenes()
nearestAnnotation <- matchGenes(x = regions, subject = genes)

## Get the region coverage
geneRegionCov <- getRegionCoverage(fullCov=fullCov, regions=regions,
	targetSize = 4e+07, totalMapped=colData(rse)$reads_aligned, verbose = FALSE)
geneRegionCovMeans = lapply(geneRegionCov, function(x) {
	cat(".")
	sapply(tIndexes, function(ii) rowMeans(x[,ii]))
})

## Plot region coverage for the longest DERs 
pdf('DERs_notingene.pdf', h = 5, w = 7)
plotRegionCoverage(regions=regions, 
	regionCoverage=geneRegionCovMeans,
	groupInfo= factor(unique(rse$group)), colors = trop, 
	nearestAnnotation=nearestAnnotation,
	annotatedRegions=annoRegs,
	ask=FALSE,	 whichRegions=1:10, verbose=FALSE, 
	txdb = TxDb.Hsapiens.UCSC.hg38.knownGene)
dev.off()

## Example of DER outside of any exonic region
plotRegionCoverage(regions=regions, 
	regionCoverage=geneRegionCovMeans,
	groupInfo= factor(unique(rse$group)), colors = trop, 
	nearestAnnotation=nearestAnnotation,
	annotatedRegions=annoRegs,
	ask=FALSE,	 whichRegions=6, verbose=FALSE, 
	txdb = TxDb.Hsapiens.UCSC.hg38.knownGene)

```
There are `r length(regions_counts_notinANYgene[regions_counts_notinANYgene$q.mod<0.05])` DERs that do not overlap with any known genes.

These range in size from `r min(width(regions_counts_notinANYgene))` bps to `r max(width(regions_counts_notinANYgene))` bps in length, with a median length of `r median(width(regions_counts_notinANYgene))` bps.

# Compare DERs to known genomic annotation
Finally, to compare the DER analysis to gene-level results, we must assign each DER to its nearest gene, creating the output file: "AnnotatedDERs.Rdata".

```{r annotated-ders}
## Assign regions to annotate with their genomic position
regions = regions_counts

## Obtain annotation information
ensemblAnno = annotateRegions(regions,gs)
countTable = ensemblAnno$countTable

## Assign overlap between regions and genomic annotation
dA = distanceToNearest(regions_counts, genes)
regions$nearestSymbol = genes$Entrez[subjectHits(dA)]
regions$distToGene = mcols(dA)$distance
mcols(regions) = cbind(mcols(regions), countTable)

## Only include regions that are within genes to compare to gene-level findings
regions_exons <- regions[regions$exon==1]
regions_exons_distToGene <- regions_exons[regions_exons$distToGene==0]

## Extract p-value and nearest gene information to be used in comparison to gene-level analysis
annotatedDERs <- regions_exons_distToGene$p.mod
names(annotatedDERs) <- regions_exons_distToGene$nearestSymbol

## Save file for use in comparison analysis
save(annotatedDERs, file="AnnotatedDERs.Rdata")

```

# Reproducibility

This analysis report was made possible thanks to:

* R `r citep(bib[['R']])`
* `r Biocpkg('BiocParallel')` `r citep(bib[['BiocParallel']])`
* `r Biocpkg('BiocStyle')` `r citep(bib[['BiocStyle']])`
* `r Biocpkg('derfinder')` `r citep(bib[['derfinder']])`
* `r CRANpkg('devtools')` `r citep(bib[['devtools']])`
* `r CRANpkg('downloader')` `r citep(bib[['downloader']])`
* `r Biocpkg('edgeR')` `r citep(bib[['edgeR']])`
* `r Biocpkg('GenomicRanges')` `r citep(bib[['GenomicRanges']])`
* `r CRANpkg('knitcitations')` `r citep(bib[['knitcitations']])`
* `r CRANpkg('matrixStats')` `r citep(bib[['matrixStats']])`
* `r Githubpkg('leekgroup/recount')` `r citep(bib[['recount']])`
* `r CRANpkg('rmarkdown')` `r citep(bib[['rmarkdown']])`
* `r Githubpkg('alyssafrazee/RSkittleBrewer')` `r citep(bib[['RSkittleBrewer']])`
* `r Biocpkg('SummarizedExperiment')` `r citep(bib[['SummarizedExperiment']])`
* `r Biocpkg('topGO')` `r citep(bib[['topGO']])`
* `r Biocpkg('limma')` `r citep(bib[['voom']])`

[Bibliography file](recount_DER_SRP032789.bib)

```{r bibliography, results='asis', echo=FALSE, warning = FALSE}
## Print bibliography
bibliography()
```


```{r reproducibility}
## Time spent creating this report:
diff(c(timestart, Sys.time()))

## Date this report was generated
message(Sys.time())

## Reproducibility info
options(width = 120)
devtools::session_info()
```
